lib/ansible/playbook/play.py

```diff
@@ -107,6 +108,8 @@ def load(data, variable_manager=None, loader=None, vars=None):
             if data['hosts'] is None or all(host is None for host in data['hosts']):
                 raise AnsibleParserError("Hosts list cannot be empty - please check your playbook")
             if isinstance(data['hosts'], list):
+                if not all(type(host) is not AnsibleMapping for host in data['hosts']):
```

Comment:
This fix is on the right track, but it is too specific. This should use `module_utils.common.collections.is_sequence` to verify it is the correct type.

Reply:
Thanks for the review. Unforutnantly is_sequence does not seem to work here, I tried with the current test case and it gives the following to this print:
`        for host in hosts_field:
            print("host {}, host type {}, host is sequence {}".format(host, type(host), is_sequence(host)))`
output:
`host {'test': 'data'}, host type <class 'ansible.parsing.yaml.objects.AnsibleMapping'>, host is sequence False`

Reply:
I've been looking through the code as well as poking at `hosts: ` and breaking things in various ways. It seems like the whole conditional here needs to be reworked and simplified. It is doing too much, which makes fixing this cleanly more challenging.

Reply:
The validation behavior changes quite a bit based on whether or not `name` was defined in the playbook. That should be handled outside of this method.

Reply:
Yes it seams a bit complicated. I pushed a solution to this, but I am still not sure whether is_sequence can be used here

Reply:
@samdoran Is my new patch fitting?

Reply:
Closer but I'm still not sure it's all the way there. Checking for `AnsibleMapping` specifically is a bit narrow. Instead we should check for valid types.